### ISSUE TITLE
Background timer working on Android, but not iOS

### ISSUE THREAD

**Issue by DigohD** [ocetnik/react-native-background-timer#54]
When running the same app on Android and iOS, it works perfectly on Android, but on iOS the timer pauses when either pressing the home button or the sleep button. What might I be doing wrong?

**Comment by cinder92** [comment]
did you solve this issue?

**Comment by DigohD** [comment]
No I'm still clueless. It simply does not work.

**Comment by Metroninja** [comment]
I'm having iOS issues as well, the Background time fires without issue on the Simulator but on a device (iphone 7s ios 10.3) the `BackgroundTimer.setInterval` will fire ~5% of the time.  In testing I have said interval set to 10 seconds (10000ms) but on the rare occasion when the interval does fire, it kicks off a fetch method who's promise never returns/resolves until I open the app back up.  Again everything works without issue in the iOS simulator.

Not sure what to do here, I'm currently only starting up the BackgroundTask by monitoring the RN AppState, once it's in the background I kick off the BackgroundTimer.  I'm on RN 0.47.1 and using  this library `1.2.0`

**Comment by DigohD** [comment]
I can't recall my timer firing even once in background mode (iPhone 6s). There are apparently some restrictions when it comes to Apple and background code execution that I've heard from people on stack overflow, but if your timer fires 5% of the time it sounds even weirder. 

I've given up on the timers for now and will try to use local notifications instead in order to pass information to the user while not having the app open.

**Comment by Metroninja** [comment]
@DigohD - that's actually what I'm attempting to do here, is use local notifications.  As soon as the app moves to the background almost all code stops executing after a few seconds making it incredibly difficult to fire local notifications on a given event since the code more or less is dead on an iOS device.  The only way I can see this working is with remote notifications if the BackgroundTimer doesn't work.

**Comment by martijnplaat** [comment]
I have the same problems. On Android process still runs when leaving the app open but on the background, but same scenario on iOS doenst work. When I return back to the app the process continue..

Also using  RN 0.47.1 and using this library v1.2.0

**Comment by perfecten** [comment]
I have the same problems on iPhone devices.

but it is ok on the Simulator.

RN 0.45
this library v1.2.0

**Comment by Metroninja** [comment]
So this library is probably never going to work unless they implement a feature similar to `react-native-background-geolocation`.  For now I'm using this library on Android, and for iOS I'm running the BackgroundGeolocation with an undocumented watcher `BackgroundGeolocation.on('heartbeat', this.heartbeat.bind(this));` and in the config ```
```
        preventSuspend: true,
        heartbeatInterval: 15,
```
 so every 15 seconds I'm doing my background tasks.

**Comment by y2k-shubham** [comment]
I'm afraid I'm facing even more grave issue: timers won't fire in iOS at all, even when the app is in the foreground.

I've just created a sample app (react-native init) and put BackgroundTimer.setInterval() inside componentDidMount(), which works perfectly fine on Android (even in background unless app is killed from recent apps)
But no timers are fired whatsoever in iOS even when the app is up and running.

react-native v0.47.2
react-native-background-timer v1.2.0
simulator iPhone 6 - iOS 10.3 (14E8301)

**Comment by martijnplaat** [comment]
thanks

**Comment by orisystech** [comment]
I have the same problem since I upgrade react native to 0.47.2 and this plugin to 1.2.0. the timer work if app is in foreground but not in background.

**Comment by Jlexyc** [comment]
It worked before for me... but i've got this bug in regression from QAs... Background timer keep working when app in forground but stops when in background.

Strange. Going to check.

**Comment by Metroninja** [comment]
More or less this library won't really ever work again - basically it's going to be less and less reliable as versions of android/ios continue to grow.  Explained very well below

https://github.com/transistorsoft/react-native-background-geolocation/issues/311#event-1220291540

**Comment by martijnplaat** [comment]
Currently I use react-native-background-geolocation heartbeat for iOS and react-native-background-timer for Android to send the GPS coordinates to a remote server in background mode.

However, reading the comments here (and on [transistorsoft/react-native-background-geolocation#311 (comment)](https://github.com/transistorsoft/react-native-background-geolocation/issues/311#event-1220291540)) I have the feeling my current implementation is not really future proof.

Am I correct? What is a future proof solution with React Native to automatically send GPS coordinates in background mode? Could I better invest in a solution using Native Modules?

**Comment by wbattel4607** [comment]
Yeah, this is going to be a problem...

**Comment by christocracy** [comment]
[Can't I just use react-native-background-timer](https://youtu.be/q6UMmLMnLl8?t=660)?

No.

**Comment by ParthBarot-BoTreeConsulting** [comment]
I believe Apple does not support running infinite/long tasks in background mode, even this lib does not work as when the app goes in background mode the iOS seems to stop the background thread.

**Comment by christocracy** [comment]
@ParthBarot-BoTreeConsulting Precisely.  This has been known for years in iOS.  Once-upon-a-time, Apple allowed these tasks to last 10 minutes.  They've shrunk it down to exactly 180 seconds since iOS 6 or 7 due to app developers abusing this mechanism, killing their customers' device battery.

**Comment by Jlexyc** [comment]
Fix with workaround added. Timer will work while iOS allows to run background task.

**Comment by christocracy** [comment]
200 seconds **max** in background.  That's all iOS will allow a `UIBackgroundTask` to live before forcefully executing your [expirationHandler](https://github.com/Jlexyc/react-native-background-timer/blob/cb90c5c4141c703f9a7b8ccd0b593cb200c36b0d/ios/RNBackgroundTimer.m#L25).

**Comment by christocracy** [comment]
Correction:  iOS 11.1 now seems to have decreased the the background-time to ~40 seconds.

**Comment by ocetnik** [comment]
@Jlexyc #66 merged and v2.0.0 released
